File: apps/desktop_flutter/lib/features/agents/models/agent_session.dart
import '../../../app/core/utils/json_parsing.dart';

enum AgentSessionStatus {
  starting('starting'),
  working('working'),
  idle('idle'),
  resumable('resumable'),
  closed('closed'),

  /// OPC-M1-4: Persisted error state — replaces the old in-memory 5s setTimeout sentinel.
  /// Cleared only on an explicit user action (new prompt / resume).
  error('error');

  final String wireValue;
  const AgentSessionStatus(this.wireValue);

  static AgentSessionStatus fromWire(String s) =>
      AgentSessionStatus.values.firstWhere(
        (k) => k.wireValue == s,
        orElse: () => AgentSessionStatus.closed,
      );
}

/// The four permission modes available for an agent session.
enum PermissionMode {
  /// Wait for user confirmation on every tool call.
  defaultMode('default'),

  /// Auto-accept write/edit tools; wait on others.
  acceptEdits('acceptEdits'),

  /// Auto-deny all tool calls (plan-only mode).
  plan('plan'),

  /// Auto-accept all tool calls without user confirmation.
  bypassPermissions('bypassPermissions');

  final String wireValue;
  const PermissionMode(this.wireValue);

  static PermissionMode fromWire(String? s) => PermissionMode.values.firstWhere(
        (m) => m.wireValue == s,
        orElse: () => PermissionMode.defaultMode,
      );

  String get displayLabel {
    switch (this) {
      case PermissionMode.defaultMode:
        return 'Default';
      case PermissionMode.acceptEdits:
        return 'Accept Edits';
      case PermissionMode.plan:
        return 'Plan Only';
      case PermissionMode.bypassPermissions:
        return 'Bypass All';
    }
  }

  String get description {
    switch (this) {
      case PermissionMode.defaultMode:
        return 'Prompt for every tool call.';
      case PermissionMode.acceptEdits:
        return 'Auto-accept write/edit tools; prompt for others.';
      case PermissionMode.plan:
        return 'Deny all tools — plan mode only.';
      case PermissionMode.bypassPermissions:
        return 'Auto-accept all tools without confirmation.';
    }
  }
}

class AgentSession {
  const AgentSession({
    required this.id,
    this.taskId,
    required this.agentId,
    required this.status,
    this.statusMessage,
    this.sessionToken,
    required this.cwd,
    required this.name,
    this.projectId,
    this.providerId,
    this.modelId,
    this.permissionMode = PermissionMode.defaultMode,
    this.thinkingBudget,
    this.fastMode = false,
    this.lastPreview,
    this.lastActivityAt,
    this.archivedAt,
    required this.createdAt,
    required this.updatedAt,
    this.parentId,
    this.sdkSessionId,
    this.anthropicAccountId,
    this.isSystem = false,
    this.category = 'chat',
    this.worktreeName,
    this.worktreePath,
    this.worktreeBranch,
  });

  final String id;
  final String? taskId;
  final String agentId;
  final AgentSessionStatus status;

  /// OPC-M1-4: Human-readable error message when status=error. Null otherwise.
  final String? statusMessage;
  final String? sessionToken;
  final String cwd;
  final String name;
  final String? projectId;
  final String? providerId;
  final String? modelId;
  final PermissionMode permissionMode;

  /// Reasoning budget in tokens (null = off). Only used when the model supports thinking.
  final int? thinkingBudget;

  /// When true, ask the SDK to use fast-mode (lower latency, less thorough).
  final bool fastMode;
  final String? lastPreview;
  final DateTime? lastActivityAt;
  final DateTime? archivedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  /// #743 — Local id of the parent session when this is a delegated subagent session.
  /// Null for top-level interactive sessions.
  final String? parentId;

  bool get isArchived => archivedAt != null;

  /// True when this session was spawned by a parent (delegated subagent).
  bool get isChildSession => parentId != null;

  /// #861 — the engine (opencode SDK) session id backing this local session,
  /// when known. Lets a delegated Task card resolve to the EXISTING local
  /// child session (persisted transcript) instead of refetching from the
  /// engine.
  final String? sdkSessionId;

  /// Anthropic account id this session is routed to (dual-account feature).
  /// Null means the app/profile default; updated live on spillover.
  final String? anthropicAccountId;

  /// #1090 — mirrors the server's `agent_sessions.is_system` column: true for
  /// background/system sessions (skill-extract, scheduler loops, memory
  /// consolidation) that must never surface in the normal chats list.
  final bool isSystem;

  /// #1090 — mirrors the server's `agent_sessions.category` column
  /// ('chat' | 'scheduled' | 'self_improvement'). Stamped at creation; drives
  /// which `AgentSessionScope` a session belongs to.
  final String category;

  /// OCU-17/18 (#1058/#1059) — when this session runs in an isolated git
  /// worktree, its name/directory/branch. All null for a normal session.
  final String? worktreeName;
  final String? worktreePath;
  final String? worktreeBranch;

  /// True when this session is running in an isolated git worktree.
  bool get isIsolatedWorktree => worktreePath != null;

  factory AgentSession.fromJson(Map<String, dynamic> json) {
    // Accept `agent_id` (new) or fall back to `agent_kind` (legacy) for one
    // release, normalising the wire value to the canonical agentId string.
    final agentId = asString(json['agent_id']) ??
        asString(json['agentId']) ??
        asString(json['agent_kind']) ??
        asString(json['agentKind']) ??
        'claude-code';
    return AgentSession(
      id: asString(json['id']) ?? '',
      taskId: asString(json['taskId']),
      agentId: agentId,
      status: AgentSessionStatus.fromWire(asString(json['status']) ?? ''),
      statusMessage: asString(json['statusMessage']),
      sessionToken: asString(json['sessionToken']),
      cwd: asString(json['cwd']) ?? '',
      name: asString(json['name']) ?? '',
      projectId: asString(json['projectId']),
      providerId: asString(json['providerId']),
      modelId: asString(json['modelId']),
      permissionMode: PermissionMode.fromWire(asString(json['permissionMode'])),
      thinkingBudget: json['thinkingBudget'] as int?,
      fastMode: json['fastMode'] as bool? ?? false,
      lastPreview: asString(json['lastPreview']),
      lastActivityAt: _parseDateTime(asString(json['lastActivityAt'])),
      archivedAt: _parseDateTime(asString(json['archivedAt'])),
      createdAt: _parseDateTime(asString(json['createdAt'])) ?? _epoch,
      updatedAt: _parseDateTime(asString(json['updatedAt'])) ?? _epoch,
      parentId: asString(json['parentSessionId']) ?? asString(json['parentId']),
      sdkSessionId:
          asString(json['sdkSessionId']) ?? asString(json['sdk_session_id']),
      anthropicAccountId: asString(json['anthropicAccountId']),
      isSystem: json['isSystem'] as bool? ?? false,
      category: asString(json['category']) ?? 'chat',
      worktreeName: asString(json['worktreeName']),
      worktreePath: asString(json['worktreePath']),
      worktreeBranch: asString(json['worktreeBranch']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      if (taskId != null) 'taskId': taskId,
      'agent_id': agentId,
      'status': status.wireValue,
      if (statusMessage != null) 'statusMessage': statusMessage,
      if (sessionToken != null) 'sessionToken': sessionToken,
      'cwd': cwd,
      'name': name,
      if (projectId != null) 'projectId': projectId,
      if (providerId != null) 'providerId': providerId,
      if (modelId != null) 'modelId': modelId,
      'permissionMode': permissionMode.wireValue,
      if (thinkingBudget != null) 'thinkingBudget': thinkingBudget,
      'fastMode': fastMode,
      if (lastPreview != null) 'lastPreview': lastPreview,
      if (lastActivityAt != null)
        'lastActivityAt': lastActivityAt!.toUtc().toIso8601String(),
      'archivedAt': archivedAt?.toUtc().toIso8601String(),
      'createdAt': createdAt.toUtc().toIso8601String(),
      'updatedAt': updatedAt.toUtc().toIso8601String(),
      if (parentId != null) 'parentSessionId': parentId,
      if (sdkSessionId != null) 'sdkSessionId': sdkSessionId,
      if (anthropicAccountId != null) 'anthropicAccountId': anthropicAccountId,
      if (worktreeName != null) 'worktreeName': worktreeName,
      if (worktreePath != null) 'worktreePath': worktreePath,
      if (worktreeBranch != null) 'worktreeBranch': worktreeBranch,
    };
  }

  AgentSession copyWith({
    String? id,
    Object? taskId = _sentinel,
    String? agentId,
    AgentSessionStatus? status,
    Object? statusMessage = _sentinel,
    Object? sessionToken = _sentinel,
    String? cwd,
    String? name,
    Object? providerId = _sentinel,
    Object? modelId = _sentinel,
    PermissionMode? permissionMode,
    Object? thinkingBudget = _sentinel,
    bool? fastMode,
    Object? lastPreview = _sentinel,
    Object? lastActivityAt = _sentinel,
    Object? archivedAt = _sentinel,
    DateTime? createdAt,
    DateTime? updatedAt,
    Object? parentId = _sentinel,
    Object? sdkSessionId = _sentinel,
    Object? anthropicAccountId = _sentinel,
    Object? worktreeName = _sentinel,
    Object? worktreePath = _sentinel,
    Object? worktreeBranch = _sentinel,
  }) {
    return AgentSession(
      id: id ?? this.id,
      taskId: taskId == _sentinel ? this.taskId : taskId as String?,
      agentId: agentId ?? this.agentId,
      status: status ?? this.status,
      statusMessage: statusMessage == _sentinel
          ? this.statusMessage
          : statusMessage as String?,
      sessionToken: sessionToken == _sentinel
          ? this.sessionToken
          : sessionToken as String?,
      cwd: cwd ?? this.cwd,
      name: name ?? this.name,
      projectId: projectId,
      providerId:
          providerId == _sentinel ? this.providerId : providerId as String?,
      modelId: modelId == _sentinel ? this.modelId : modelId as String?,
      permissionMode: permissionMode ?? this.permissionMode,
      thinkingBudget: thinkingBudget == _sentinel
          ? this.thinkingBudget
          : thinkingBudget as int?,
      fastMode: fastMode ?? this.fastMode,
      lastPreview:
          lastPreview == _sentinel ? this.lastPreview : lastPreview as String?,
      lastActivityAt: lastActivityAt == _sentinel
          ? this.lastActivityAt
          : lastActivityAt as DateTime?,
      archivedAt:
          archivedAt == _sentinel ? this.archivedAt : archivedAt as DateTime?,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      parentId: parentId == _sentinel ? this.parentId : parentId as String?,
      sdkSessionId: sdkSessionId == _sentinel
          ? this.sdkSessionId
          : sdkSessionId as String?,
      anthropicAccountId: anthropicAccountId == _sentinel
          ? this.anthropicAccountId
          : anthropicAccountId as String?,
      isSystem: isSystem,
      category: category,
      worktreeName: worktreeName == _sentinel
          ? this.worktreeName
          : worktreeName as String?,
      worktreePath: worktreePath == _sentinel
          ? this.worktreePath
          : worktreePath as String?,
      worktreeBranch: worktreeBranch == _sentinel
          ? this.worktreeBranch
          : worktreeBranch as String?,
    );
  }
}

// Sentinel used for nullable copyWith parameters.
const Object _sentinel = Object();

final DateTime _epoch = DateTime.fromMillisecondsSinceEpoch(0);

DateTime? _parseDateTime(String? value) {
  if (value == null) return null;
  final parsed = DateTime.tryParse(value);
  if (parsed == null) return null;
  return parsed;
}

File: apps/api_server/src/models/agent_session.ts
export type AgentKind = 'claude-code' | 'codex';
export type AgentSessionStatus = 'starting' | 'working' | 'idle' | 'resumable' | 'closed' | 'error';

export type PermissionMode = 'default' | 'acceptEdits' | 'plan' | 'bypassPermissions';

export const PERMISSION_MODES: PermissionMode[] = ['default', 'acceptEdits', 'plan', 'bypassPermissions'];

/**
 * USO A1 (#1024) — the `scope` query param on GET /agent-sessions selects which
 * slice of the session list to return. 'chats' is the default and preserves the
 * pre-USO behavior (interactive chats only, is_system=0).
 */
export type SessionScope = 'chats' | 'scheduled' | 'self_improvement';
export const SESSION_SCOPES: SessionScope[] = ['chats', 'scheduled', 'self_improvement'];

/**
 * USO B1 (#1028) — persisted classification stamped on every agent_sessions row
 * (column `category`, default 'chat'). Drives the USO scope filters:
 *   'chat'            → interactive chat sessions (the default Chats view)
 *   'scheduled'       → runs tied to a scheduled task
 *   'self_improvement'→ curator/skill/memory background runs
 */
export type SessionCategory = 'chat' | 'scheduled' | 'self_improvement';
export const SESSION_CATEGORIES: SessionCategory[] = ['chat', 'scheduled', 'self_improvement'];

export interface AgentSession {
  id: string;
  taskId: string | null;
  taskTitle: string | null;
  agentKind: AgentKind;
  status: AgentSessionStatus;
  /** Human-readable error message when status='error'. Null otherwise. */
  statusMessage: string | null;
  /**
   * C1 — MCP role name (e.g. "church-admin") resolved from .mcp-roles/<role>.mcp.json
   * at session-create time. Null when no role was requested. The resolved
   * allowlist is stored alongside; see mcpAllowedToolsJson.
   */
  mcpRole: string | null;
  /**
   * C1 — JSON-serialised per-server allowedTools map from the resolved MCP role.
   * Shape: Record<serverName, string[]>. Null when mcpRole is null.
   * Used by the WS gateway (future: init-time gate enforcement).
   */
  mcpAllowedToolsJson: string | null;
  /**
   * @deprecated Use sdkSessionId for resume continuity (OPC-M1-5).
   * Retained for backward compatibility — do not remove the column.
   */
  sessionToken: string | null;
  /**
   * OPC-M1-5 — The Opencode SDK session id (e.g. "ses_abc123").
   * Set at session create time; used by resume() to re-attach to the existing
   * SDK conversation rather than creating a fresh session. Null for sessions
   * created before this migration.
   */
  sdkSessionId: string | null;
  cwd: string;
  name: string;
  projectId: string | null;
  providerId: string | null;
  modelId: string | null;
  agentMode: string | null;
  permissionMode: PermissionMode;
  /** Reasoning budget in tokens (null = off). Only applied when the model supports extended thinking. */
  thinkingBudget: number | null;
  /** When true, ask the SDK to use fast-mode (lower latency, less thorough). */
  fastMode: boolean;
  lastPreview: string | null;
  lastActivityAt: string | null;
  archivedAt: string | null;
  createdAt: string;
  updatedAt: string;
  /**
   * Agent-loop tracking: the `agent_scheduled_tasks.id` that triggered this
   * session. Null for interactive sessions.
   */
  scheduledTaskId: string | null;
  /**
   * #743 — Local id of the parent agent_sessions row that spawned this session
   * via the `task` tool delegation. Null for top-level interactive sessions.
   */
  parentSessionId: string | null;
  /**
   * #747 — When true, this is a background/system session (skill-extract,
   * skill-refine-judge, scheduler background-loop, memory consolidation) and
   * must NOT appear in the normal session list or agent picker. Child sessions
   * (#743, parent_session_id NOT NULL) that are delegated subagent tasks remain
   * user-visible (isSystem = false).
   */
  isSystem: boolean;
  /**
   * Dual Anthropic accounts (Task D) — the account id this session's Anthropic
   * requests are routed to (mirrors the routing map in anthropic-accounts.json).
   * Null = engine default. Updated in place on rate-limit spillover.
   */
  anthropicAccountId: string | null;
  /** Owner used for owner-scoped agent memory. Null means instance-global only. */
  ownerUserId: number | null;
  /** Delegation nesting depth for rhythm_delegate-created runs. Root sessions default to 0. */
  delegationDepth: number;
  /**
   * USO B1 (#1028) — session classification driving the USO scope filters.
   * Stamped at creation ('scheduled' when scheduledTaskId is set, 'self_improvement'
   * when the caller requests it, else 'chat'). Legacy rows are derived at
   * migration time (scheduled_task_id NOT NULL → 'scheduled', else 'chat').
   */
  category: SessionCategory;
  /**
   * OCU-17 (#1058) — when this session runs in an isolated git worktree, the
   * worktree's name / directory / branch (so the UI and hard-delete cleanup can
   * act on it). All null for a normal (non-isolated) session.
   */
  worktreeName: string | null;
  worktreePath: string | null;
  worktreeBranch: string | null;
}

export interface UpdateAgentSessionDto {
  name?: string;
  providerId?: string | null;
  modelId?: string | null;
  agentMode?: string | null;
}

export interface AgentSessionMessage {
  id: number;
  sessionId: string;
  role: 'output' | 'input' | 'system';
  rawText: string;
  strippedText: string;
  createdAt: string;
  /** SDK message id (null for legacy rows). */
  sdkMessageId: string | null;
  /** Raw JSON string of ordered part array (null for legacy rows). */
  partsJson: string | null;
  /** Raw JSON string of token usage object (null for legacy rows). */
  tokensJson: string | null;
  /** Message cost in USD (null for legacy rows). */
  cost: number | null;
}

/**
 * AgentSessionMessage with parts/tokens deserialized.
 * Returned by listBySessionStructured() — what the GET /messages endpoint sends.
 */
export interface StructuredAgentSessionMessage extends Omit<AgentSessionMessage, 'partsJson' | 'tokensJson'> {
  /** Parsed part array. Legacy rows get a synthetic [{type:'text', text:rawText}]. */
  parts: unknown[];
  /** Parsed token usage object, or null for legacy rows. */
  tokens: Record<string, unknown> | null;
}

export interface CreateAgentSessionDto {
  agentKind: AgentKind;
  taskId: string | null;
  taskTitle?: string | null;
  cwd: string;
  name: string;
  projectId?: string | null;
  /** C1 — optional MCP role slug (e.g. "church-admin"). Null/undefined = no scoping. */
  mcpRole?: string | null;
  /**
   * C1 — Resolved per-server allowedTools from the role file.
   * Shape: Record<serverName, string[]>. Null when no role was requested.
   */
  mcpAllowedToolsJson?: string | null;
  /**
   * Agent-loop tracking: the `agent_scheduled_tasks` row that triggered this
   * session. Null for interactive sessions. Persisted so the CHATS list can
   * surface scheduler-originated runs with their source task.
   */
  scheduledTaskId?: string | null;
  /** #747 — Mark this session as a background system session (curator, scheduler, memory). */
  isSystem?: boolean;
  /** Dual Anthropic accounts (Task D) — resolved account id (body → profile → store default). Null = engine default. */
  anthropicAccountId?: string | null;
  /** Owner used for owner-scoped agent memory. Null means instance-global only. */
  ownerUserId?: number | null;
  /** Delegation nesting depth. Root sessions default to 0. */
  delegationDepth?: number;
  /**
   * USO B1 (#1028) — explicit session category. When omitted it is derived at
   * insert time: 'scheduled' if scheduledTaskId is set, otherwise 'chat'. Pass
   * 'self_improvement' for curator/skill/memory background runs.
   */
  category?: SessionCategory;
}
